infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java

```diff
@@ -390,7 +408,25 @@ public void resetServer(CorfuPayloadMsg<SequencerTailsRecoveryMsg> msg,
 
             // Clear the existing map as it could have been populated by an earlier reset.
             streamTailToGlobalTailMap.clear();
-            streamTailToGlobalTailMap.putAll(streamTails);
+
+            // Set tail for every stream
+            // TODO (Anny): assess cost of computation vs. sending stream tails as well on reset
+            for(Map.Entry<UUID, StreamAddressSpace> streamAddressSpace : streamsAddressSpaceMap.entrySet()) {
+                Long streamTail;
+                // If no address is present for this stream, the tail is given by the trim mark (last trimmed address)
+                if(streamAddressSpace.getValue().getAddressCount() == 0) {
```

Comment:
you should move this if-else condition inside `StreamAddressSpace` class. Just provide a method which will contain this if-else logic, the method name could be: `getStreamTail()`.
P.S. always try not to expose the internal state if it's possible like in this case. (Encapsulation, Single Responsibility principle)

Reply:
done.